gcp_variant_transforms/libs/conflicts_reporter.py

```diff
@@ -0,0 +1,139 @@
+
+from gcp_variant_transforms.beam_io.vcf_header_io import VcfHeader  # pylint: disable=unused-import
+from gcp_variant_transforms.beam_io.vcf_header_io import VcfParserHeaderKeyConstants
+from gcp_variant_transforms.transforms.merge_header_definitions import Definition  # pylint: disable=unused-import
+from gcp_variant_transforms.transforms.merge_header_definitions import VcfHeaderDefinitions  # pylint: disable=unused-import
+
+
+_HEADER_LINE = 'ID;Conflicts;Proposed Resolution\n'
+_NO_CONFLICTS_MESSAGE = 'No conflicts found.'
+_NO_SOLUTION_MESSAGE = 'Not resolved'
+
+
+def generate_conflicts_report(file_path,
+                              header_definitions,
+                              resolved_headers=None):
+  # type: (str, VcfHeaderDefinitions, VcfHeader) -> None
+  """Generates a report.
+
+  Combines the conflicts extracted from ``header_definitions`` and their
+  resolutions from ``resolved_headers`` to generate a conflicts report.
+  ``file_path`` specifies the location where the conflicts report is saved.
+  """
+  resolved_headers = resolved_headers or VcfHeader()
+  content_lines = []
+  content_lines.extend(_generate_contents(
+      _extract_conflicts(header_definitions.formats), resolved_headers.formats))
+  content_lines.extend(_generate_contents(
+      _extract_conflicts(header_definitions.infos), resolved_headers.infos))
+  _write_to_report(content_lines, file_path)
+
+
+def _extract_conflicts(
+    definitions  # type: Dict[str, Dict[Definition, List[str]]]
+):
+  # type: (...) -> Dict[str, Dict[Definition, List[str]]]
+  """Extracts the fields that have conflicted definitions.
+
+  Returns:
+    A dictionary that maps field id with conflicted definitions to a dictionary
+    which maps ``Definition`` to a list of file names.
+  """
+  # len(v) > 1 means there are conflicted definitions for this field.
+  return dict([(k, v) for k, v in definitions.items() if len(v) > 1])
+
+
+def _generate_contents(
+    conflicts,  # type: Dict[str, Dict[Definition, List[str]]]
+    resolved_headers  # type: Dict[str, Dict[str, Union[str, int]]
+):
+  # type: (...) -> List[str]
+  """Generates the report contents.
+
+  The conflicted definitions, the file names and the resolutions are included
+  in the contents.
+  Output example:
+  (NS; num=1 type=Float in ['file1','file2'] num=1 type=Integer in ['file3'];
+  num=1 type=Float)
+  """
+  contents = []
+  for field_id, definitions_to_files_map in conflicts.iteritems():
+    row = [
+        field_id,
+        _extract_definitions_and_file_names(definitions_to_files_map),
+        _extract_resolution(resolved_headers, field_id)
+    ]
+    contents.append(';'.join(row))
+  return contents
+
+
+def _extract_definitions_and_file_names(definition_to_files_map):
+  # type: (Dict[Definition, List[str]]) -> str
+  """Extracts the definitions and related file names.
+
+  Output example:
+  num=1 type=Float in ['file1','file2'] num=1 type=Integer in ['file3']
```

Comment:
nit: add a comma in between?

`num=1 type=Float in ['file1','file2'], num=1 type=Integer in ['file3']`

Reply:
Thanks, better!